File: solutions/ocaml/049-add-binary.ml
(*
  Problem 49: Add Binary (LeetCode 67)
  Difficulty: Easy
  Language: OCaml
*)
let add_binary a b =
  let rec go i j carry =
    if i < 0 && j < 0 && carry = 0 then ""
    else
      let da = if i >= 0 then Char.code a.[i] - 48 else 0 in
      let db = if j >= 0 then Char.code b.[j] - 48 else 0 in
      let s = da + db + carry in
      go (i-1) (j-1) (s / 2) ^ string_of_int (s mod 2)
  in go (String.length a - 1) (String.length b - 1) 0

File: solutions/clojure/049-add-binary.clj
;; Problem 49: Add Binary (LeetCode 67)
;; Difficulty: Easy
;; Language: Clojure
;; 
(defn add-binary [a b]
  (loop [i (dec (count a)) j (dec (count b)) carry 0 acc ""]
    (if (and (neg? i) (neg? j) (zero? carry)) acc
      (let [da (if (>= i 0) (- (int (nth a i)) 48) 0)
            db (if (>= j 0) (- (int (nth b j)) 48) 0)
            s (+ da db carry)]
        (recur (dec i) (dec j) (quot s 2)
               (str (mod s 2) acc))))))
